File: librarymanager-be/src/main/java/ru/elenacherev/librarymanager/services/LangService.java
package ru.elenacherev.librarymanager.services;

import lombok.RequiredArgsConstructor;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import ru.elenacherev.librarymanager.api.dto.Edition;
import ru.elenacherev.librarymanager.api.dto.Lang;
import ru.elenacherev.librarymanager.domain.entity.LangEntity;
import ru.elenacherev.librarymanager.domain.repository.EditionRepository;
import ru.elenacherev.librarymanager.domain.repository.LangRepository;
import ru.elenacherev.librarymanager.mapper.EditionMapper;
import ru.elenacherev.librarymanager.mapper.LangMapper;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;


@RequiredArgsConstructor
@Service
public class LangService {

	private final LangRepository langRepository;
	private final EditionRepository editionRepository;

	@Transactional
	public Lang save(Lang dto) {
		return langRepository.findById(dto.getLangId())
				.map(entity -> LangMapper.map(dto, entity))
				.map(langRepository::saveAndFlush)
				.map(LangMapper::map)
				.orElse(null);
	}

	@Transactional(readOnly = true)
	public List<Lang> findAll() {
		return langRepository.findAll()
				.stream()
				.map(LangMapper::map)
				.collect(Collectors.toList());
	}

	@Transactional(readOnly = true)
	public Lang findLangById(Long langId) {
		return langRepository.findById(langId)
				.map(LangMapper::map)
				.orElse(null);
	}

	@Transactional(readOnly = true)
	public Page<Edition> findEditionsByLangId(Long langId, Pageable pageable) {
		return langRepository.findById(langId)
				.map(entity -> editionRepository.findAllByLang(entity, pageable))
				.orElse(Page.empty())
				.map(EditionMapper :: map);
	}
	
	@Transactional(readOnly = true)
	public Page<Edition> findEditionsByPublLangId(Long langId, Pageable pageable) {
		return langRepository.findById(langId)
				.map(entity -> editionRepository.findAllByPublLang(entity, pageable))
				.orElse(Page.empty())
				.map(EditionMapper :: map);
	}

	public List<Edition> updateEditionsByLangId(Long langId, List<Long> editionIds) {
		return langRepository
				.findById(langId)
				.map(langEntity-> {
					langEntity.getEditions().clear();
					langEntity.getEditions().addAll(
							editionIds.stream()
  							.map(editionRepository::getOne)
  							.collect(Collectors.toList())
  							);
					return langEntity;
				 })
				.map(langRepository::saveAndFlush)
				.map(LangEntity :: getEditions)
				.orElse(Collections.emptySet())
				.stream()
				.map(EditionMapper :: map)
				.collect(Collectors.toList());
	}

	public List<Edition> updateEditionsByPubllangId(Long publlangId, List<Long> editionIds) {
		return langRepository
				.findById(publlangId)
				.map(langEntity-> {
					langEntity.getPublLangEditions().clear();
					langEntity.getPublLangEditions().addAll(
							editionIds.stream()
  							.map(editionRepository::getOne)
  							.collect(Collectors.toList())
  							);
					return langEntity;
				 })
				.map(langRepository::saveAndFlush)
				.map(LangEntity :: getPublLangEditions)
				.orElse(Collections.emptySet())
				.stream()
				.map(EditionMapper :: map)
				.collect(Collectors.toList());
	}
}

File: librarymanager-be/src/main/java/ru/elenacherev/librarymanager/services/LangService.kt
package ru.elenacherev.librarymanager.services

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import ru.elenacherev.librarymanager.api.dto.Edition
import ru.elenacherev.librarymanager.api.dto.Lang
import ru.elenacherev.librarymanager.domain.entity.EditionEntity
import ru.elenacherev.librarymanager.domain.entity.LangEntity
import ru.elenacherev.librarymanager.domain.repository.EditionRepository
import ru.elenacherev.librarymanager.domain.repository.LangRepository
import ru.elenacherev.librarymanager.mapper.map

@Service
class LangService (
    val langRepository: LangRepository,
    val editionRepository: EditionRepository
){
    @Transactional
    fun save(langId:Long, dto: Lang): Lang {
        return langRepository
            .findById(langId)
            .map (dto:: map)
            .map{ obj: LangEntity -> langRepository.saveAndFlush(obj) }
            .map(LangEntity::map)
            .orElse(null)
    }

    @Transactional(readOnly = true)
    fun findAll(): List<Lang> {
        return langRepository
            .findAll()
            .map(LangEntity::map)
    }

    @Transactional(readOnly = true)
    fun findLangById(langId: Long): Lang {
        return langRepository
            .findById(langId)
            .map(LangEntity::map)
            .orElse(null)
    }

    @Transactional(readOnly = true)
    fun findEditionsByLangId(langId: Long, pageable: Pageable): Page<Edition> {
        return langRepository
            .findById(langId)
            .map { entity -> editionRepository.findAllByLang(entity, pageable) }
            .orElse(Page.empty())
            .map(EditionEntity::map)
    }

    @Transactional(readOnly = true)
    fun findEditionsByPublLangId(langId: Long, pageable: Pageable): Page<Edition> {
        return langRepository.findById(langId)
            .map { entity -> editionRepository.findAllByPublLang(entity, pageable) }
            .orElse(Page.empty())
            .map(EditionEntity::map)
    }

    fun updateEditionsByLangId(langId: Long, editionIds: List<Long>): List<Edition> {
        return langRepository
            .findById(langId)
            .map { langEntity ->
                langEntity.editions.clear()
                langEntity.editions +=
                    editionIds.map(editionRepository:: getOne)
                return@map langEntity
            }
            .map{ obj: LangEntity -> langRepository.saveAndFlush(obj) }
            .map{ obj: LangEntity -> obj.editions }
            .orElse(mutableSetOf())
            .map(EditionEntity::map)
    }

    fun updateEditionsByPubllangId(publlangId: Long, editionIds: List<Long>): List<Edition> {
        return langRepository
            .findById(publlangId)
            .map { langEntity ->
                langEntity.publLangEditions.clear()
                langEntity.publLangEditions += editionIds.map(editionRepository::getOne)
                return@map langEntity
            }
            .map{ obj: LangEntity -> langRepository.saveAndFlush(obj) }
            .map{ obj: LangEntity -> obj.publLangEditions }
            .orElse(mutableSetOf())
            .map(EditionEntity::map)
    }
}